service-base/src/main/java/org/eclipse/hono/service/tenant/TenantHttpEndpoint.java

```diff
@@ -0,0 +1,232 @@
+                    final String msg = tenantResult.getJsonObject(TenantConstants.FIELD_PAYLOAD).encodePrettily();
+                    response
+                            .putHeader(HttpHeaders.CONTENT_TYPE, HttpUtils.CONTENT_TYPE_JSON_UFT8)
+                            .putHeader(HttpHeaders.CONTENT_LENGTH, String.valueOf(msg.length()))
+                            .write(msg);
+                default:
+                    response.end();
+                }
+        });
+    }
+
+    private void doAddTenantJson(final RoutingContext ctx) {
+        try {
+            JsonObject payload = null;
+            if (ctx.getBody().length() > 0) {
+                payload = ctx.getBodyAsJson();
+            }
+            addTenant(ctx, payload);
+        } catch (final DecodeException e) {
+            HttpUtils.badRequest(ctx, "body does not contain a valid JSON object");
+        }
+    }
+
+    private void addTenant(final RoutingContext ctx, final JsonObject payload) {
+
+        if (payload == null) {
+            HttpUtils.badRequest(ctx, "missing body");
+        } else {
+            final Object tenantId = payload.remove(FIELD_TENANT_ID);
+            if (tenantId == null) {
+                HttpUtils.badRequest(ctx, String.format("'%s' param is required", FIELD_TENANT_ID));
+            } else if (!(tenantId instanceof String)) {
+                HttpUtils.badRequest(ctx, String.format("'%s' must be a string", FIELD_TENANT_ID));
+            } else {
+                logger.debug("adding tenant [tenant: {}, payload: {}]", tenantId, payload);
+                final HttpServerResponse response = ctx.response();
+                final JsonObject requestMsg = TenantConstants.getServiceRequestAsJson(Action.ACTION_ADD.toString(),
+                        (String) tenantId, payload);
+                doTenantAction(ctx, requestMsg, (status, tenantResult) -> {
+                    response.setStatusCode(status);
+                    switch(status) {
+                        case HttpURLConnection.HTTP_CREATED:
+                            response.putHeader(
+                                    HttpHeaders.LOCATION,
+                                    String.format("/%s/%s", TenantConstants.TENANT_ENDPOINT, tenantId));
+                        default:
+                            if (status >= 400) {
+                                setResponseBody(tenantResult, response);
+                            }
+                            response.end();
+                    }
+                });
+            }
+        }
+    }
+
+    private void doUpdateTenantJson(final RoutingContext ctx) {
+
+        try {
+            JsonObject payload = null;
+            if (ctx.getBody().length() > 0) {
+                payload = ctx.getBodyAsJson();
+            }
+            updateTenant(payload, ctx);
+        } catch (final DecodeException e) {
+            HttpUtils.badRequest(ctx, "body does not contain a valid JSON object");
+        }
+    }
```

Comment:
this is the same code as *doAddTenantJson*. Maybe we can refactor the body extraction code into a separate method?